Add precise scaling control to osu! editor

OliBomby opened the pull request:
This PR adds the scale popover tool from osu! stable to lazer. I tried to mimic the original functionality as closely as possible, and in some aspects improve it.

I also added a hotkey for the scaling tool which is `S`. Unfortunately more logical hotkeys like `Ctrl+S` and `Ctrl+Shift+S` are already taken.

The maximum value of the scale slider is determined based on the current choice of origin and axis. The maximum value will always be the maximum scaling value that still keeps the scaling objects on screen.

I also slightly changed how the rotation popover works. Now it will automatically select the text in the textbox input, so the user can instantly type the value they want after opening the tool.

![osu Game Rulesets Osu Tests_EcLev87sNt](https://github.com/ppy/osu/assets/17460441/b61bc89b-1552-4e1f-bc89-9e409302af53)

OliBomby committed:
Add simple scale tool

OliBomby committed:
select all input text on popup for an easy typing experience

OliBomby committed:
calculate max scale bounds for scale slider

OliBomby committed:
Enable scale buttons at the correct times

OliBomby committed:
add axis toggles

Walavouchey commented:
having rotate on `Ctrl`+`S` and scale on `S` will be weird... i think it'd be worth rethinking the hotkeys

bdach commented:
The hotkey 200% needs to change. This feature is not so central to the editor that it should be allowed to take up a letter without modifiers.

I'd probably do something either close to <kbd>Ctrl</kbd>-<kbd>S</kbd> (because **s**cale) or <kbd>Ctrl</kbd>-<kbd>R</kbd> (because next to the rotation binding) location-wise. Maybe something like <kbd>Ctrl</kbd>-<kbd>T</kbd>.

bdach commented on osu.Game/Screens/Edit/Compose/Components/SelectionScaleHandler.cs:
```diff
@@ -32,8 +32,26 @@ public partial class SelectionScaleHandler : Component
         /// </remarks>
         public Bindable<bool> CanScaleDiagonally { get; private set; } = new BindableBool();
 
+        /// <summary>
+        /// Whether scaling anchored by the selection origin can currently be performed.
+        /// </summary>
+        public Bindable<bool> CanScaleSelectionOrigin { get; private set; } = new BindableBool();
+
+        /// <summary>
+        /// Whether scaling anchored by the center of the playfield can currently be performed.
+        /// </summary>
+        public Bindable<bool> CanScalePlayfieldOrigin { get; private set; } = new BindableBool();
```
- These are named kinda weird (as if the "origin" was being scaled).
- `CanScaleSelectionOrigin` is weird here to me because `CanScale{X,Y,Diagonally}` already exist and my presumption would be that those _already_ are for scaling about the origin. I'm not really even sure what it achieves, the only usage (`PreciseScalePopover`) could just inline the `CanScaleX.Value || CanScaleY.Value` logic locally.
- `CanScalePlayfieldOrigin` cannot live in the base `SelectionScaleHandler` imo because that only makes sense in the specific contexts where a playfield is present. Should be moved to the derived `OsuSelectionScaleHandler` imo.

The "not being able to access these if these are moved to derived" issue can be fixed by casting. I don't see that as a huge concern compared to just putting stuff in the base impl that makes no sense for skinning, for instance.

bdach commented on osu.Game/Screens/Edit/Compose/Components/SelectionScaleHandler.cs:
```diff
@@ -32,8 +32,26 @@ public partial class SelectionScaleHandler : Component
+        /// <param name="origin">The origin from which the scale operation is performed</param>
+        /// <param name="scale">The scale to be clamped</param>
+        /// <returns>The clamped scale vector</returns>
+        public virtual Vector2 GetClampedScale(Vector2 scale, Vector2? origin = null) => scale;
```
Same for this, I don't think this belongs in the base class. Should be moved to derived.

bdach commented on osu.Game.Rulesets.Osu/Edit/PreciseScalePopover.cs:
```diff
@@ -0,0 +1,186 @@
+            };
+            selectionCentreButton.Selected.DisabledChanged += isDisabled =>
+            {
+                selectionCentreButton.TooltipText = isDisabled ? "Select more than one object to perform selection-based scaling." : string.Empty;
```
Tooltip text does not cover all cases where this can be disabled, this can also be disabled if the selection has zero effective height/width (like stacked circles):

![osu_2024-05-27_08-54-11](https://github.com/ppy/osu/assets/20418176/57d1fa4b-6fe2-4d9e-952f-b4b1abb8f356)

Not sure I'd bother at that point, explaining this particular case is likely to be too difficult to be worth it.

bdach replied:
My remark above seems to have been misunderstood, I meant that at that point attempting to provide a detailed explanation is likely counterproductive.

I'm changing the tooltip copy to be more generic instead (will explain that the operation is unavailable but won't attempt to explain exactly why).

bdach commented on osu.Game.Rulesets.Osu/Edit/PreciseScalePopover.cs:
```diff
@@ -0,0 +1,186 @@
+                        Spacing = new Vector2(4),
+                        Children = new Drawable[]
+                        {
+                            xCheckBox = new OsuCheckbox(false)
+                            {
+                                RelativeSizeAxes = Axes.X,
+                                LabelText = "X-axis",
+                                Current = { Value = true },
+                            },
+                            yCheckBox = new OsuCheckbox(false)
+                            {
+                                RelativeSizeAxes = Axes.X,
+                                LabelText = "Y-axis",
+                                Current = { Value = true },
+                            },
```
These behave strangely sometimes if a selection is only resizable along one axis:

https://github.com/ppy/osu/assets/20418176/cc6104b1-b743-418c-84ea-637ad47a145a

Disabled state should probably be hooked up to `CanScale{X,Y}`.

OliBomby committed:
Change scale hotkey to Ctrl+T

OliBomby committed:
rename CanScaleSelectionOrigin/PlayfieldOrigin to make clear its not the origin being scaled

OliBomby committed:
remove redundant CanScaleFromSelectionOrigin

OliBomby committed:
Bind axis checkbox disabled state to CanScaleX/Y

OliBomby committed:
refactor CanScaleFromPlayfieldOrigin and GetClampedScale to derived class

OliBomby committed:
fix axis checkboxes being disabled in playfield origin scale

OliBomby committed:
disable playfield centre origin when scaling slider and simplify logic

bdach committed:
Use better name if we're renaming

bdach committed:
Rename another method

bdach committed:
Simplify toolbox button enable logic

bdach committed:
Use more generic tooltip copy

bdach committed:
Simplify logic

bdach committed:
Rename method

bdach committed:
Change editor scale hotkey to Ctrl-E

Forgot that Ctrl-T was taken by the game-global toolbar already, so it
wasn't working.

bdach committed:
Merge branch 'master' into scale-tool

bdach commented:
Seems good enough. I've done a round of cleanups but nothing super worth mentioning in there. Well maybe aside the hotkey changing - <kbd>Ctrl</kbd>-<kbd>T</kbd> was already taken by the game-global toolbar and thus wasn't working so I did <kbd>Ctrl</kbd>-<kbd>E</kbd> instead. Head-canon that it's E for scalE, but that's kinda stretching and I just wanted to have it close to <kbd>Ctrl</kbd>-<kbd>R</kbd> and <kbd>Ctrl</kbd>-<kbd>S</kbd> I guess.

bdach committed:
Remove unused fields
